src/ess.jl

```diff
@@ -255,37 +255,39 @@ function ess_rhat(
 )
     # compute size of matrices (each chain may be split!)
     niter = size(chains, 1) ÷ split_chains
-    nparams = size(chains, 3)
     nchains = split_chains * size(chains, 2)
     ntotal = niter * nchains
 
+    # vector used to construct output vector types
+    params = view(chains, firstindex(chains, 1), firstindex(chains, 2), :)
+
     # do not compute estimates if there is only one sample or lag
     maxlag = min(maxlag, niter - 1)
-    maxlag > 0 || return fill(missing, nparams), fill(missing, nparams)
+    maxlag > 0 || return similar(params, Missing), similar(params, Missing)
 
     # define caches for mean and variance
-    U = typeof(zero(eltype(chains)) / 1)
     T = promote_type(eltype(chains), typeof(zero(eltype(chains)) / 1))
     chain_mean = Array{T}(undef, 1, nchains)
     chain_var = Array{T}(undef, nchains)
     samples = Array{T}(undef, niter, nchains)
 
     # compute correction factor
-    correctionfactor = (niter - 1) / niter
+    correctionfactor = (niter - 1)//niter
 
     # define cache for the computation of the autocorrelation
     esscache = build_cache(method, samples, chain_var)
 
     # define output arrays
-    ess = Vector{T}(undef, nparams)
-    rhat = Vector{T}(undef, nparams)
+    ess = similar(params, T)
+    rhat = similar(ess)
+    i0 = firstindex(ess) - 1
```

Comment:
Is `i` used anywhere in the loop below? If not it would be simpler (I think) to replace
```julia
for (i, chains_slice) in enumerate(eachslice(chains; dims=3))
```
with
```julia
for (i, chains_slice) in zip(eachindex(ess), eachslice(chains; dims=3))
```